### PR TITLE
WFCORE-915 allow concurrent registration of slave hosts

### PR DESCRIPTION
**Pull request by luck3y**
Also updates ModelControllerLock to use a bitmask internally for state. I closed the previous one by mistake :-1:

### CODE REVIEW

**Comment by luck3y** [comment]
retest this please

**Comment by luck3y** [comment]
retest this please

File: controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java [hunk 1]
```diff
@@ -52,78 +81,315 @@ boolean lock(Integer permit, long timeout, TimeUnit unit) throws InterruptedExce
...
-            } else if (permitHolder.get().equals(permit)) {
-                for (;;) {
-                    int current = getState();
-                    int next = current + 1; // increase by one
-                    if (next < 0) // overflow
-                        throw new Error("Maximum lock count exceeded");
-                    if (compareAndSetState(current, next)) {
-                        return true;
-                    }
+        protected boolean isHeldExclusively() {
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java]: Should be private. No big deal as the class is private, but I'm anal. ;)

File: controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java [hunk 2]
```diff
@@ -52,78 +81,315 @@ boolean lock(Integer permit, long timeout, TimeUnit unit) throws InterruptedExce
...
+
+            // can't exclusively lock, already in shared mode
+            if (mode == SHARED && exclusive) {
+                return -1;
+            }
+
+            // can't get a shared lock if its already exclusive
+            if (mode == EXCLUSIVE && !exclusive) {
+                return -1;
+            }
```
Comment (bstansberry) [inline controller/src/main/java/org/jboss/as/controller/ModelControllerLock.java]: Can L289-291 be removed and just use this block? I think then isHeldExclusively() can be removed as unused.

**Comment by luck3y** [comment]
retest this please

**Comment by luck3y** [comment]
retest this please

**Comment by bstansberry** [comment]
Are you planning on making further changes to the lock to restore the separate permit field, or is this ready to go?

**Comment by luck3y** [comment]
As long as @n1hility approves, this is g2g.

**Comment by luck3y** [comment]
@bstansberry @n1hility approves, so unless you have any other comments on this, it should be good to go.

**Pull request merged**